### ISSUE TITLE
[ReadSupport] Add support to read from relationships with provided type

### ISSUE THREAD

**Issue by conker84** [neo4j-contrib/neo4j-spark-connector#142]
We want read relationships as tables in the following way:

```scala
val simpleDf = sparkSession
   .read
   .format("neo4j")
   .option("url", "neo4j://localhost:7687")
   .option("relationship", "KNOWS")
   .load()
```

This will extract source/target nodes (with the keys as maps) and the relationships properties.
Example:
For the given set of nodes:

```cypher
CREATE (p:Person:Customer{ssn: 'AB12', name: 'Andrea', surname: 'Santurbano'})-[:KNOWS{since: 2020}]->(p:Person{ssn: 'CD34', name: 'Davide', surname: 'Fantuzzi'})
```

The output of the Dataframe will be the following:

sourceNode | sourceNodeLabels | endNode | endNodeLabels | type | since
-- | -- | -- | -- | -- | --
{ssn: ‘AB12’} | [Person, Customer] | {ssn: ‘CD34’} | [Person] | KNOWS | 2020

*N.b.* As Spark supports only maps with the same value types (i.e. Map<String, String>, Map<String, Integer>, and so on) we will coerce all to Map<String, String>. This feature will support push-down filtering in order to optimize the data moved from Neo4j to Spark

**Comment by moxious** [comment]
Warning -- neo4j doesn't yet have secondary relationship type indexes.  Unless you require the user to specify the source and target labels, this query will perform very, very badly.

This is quite bad:

```
MATCH (n)-[r:KNOWS]->(m) RETURN r
```

(It scans every node in the entire database)

This is quite doable:

```
MATCH (p:Person)-[r:KNOWS]->(m) return r;
```

Recommend not allowing people to do this without source or target label constraints.   If you have say 2,000 knows relationships in a graph of billions of nodes, this will not work well.

**Comment by conker84** [comment]
closed via e689a29
